File: backend/AI_Generation_Service/src/config/rabbitmq.js
/**
 * RabbitMQ publisher for asynchronous AI generation tasks.
 */
import amqp from 'amqplib';
import { randomUUID } from 'node:crypto';

const RABBITMQ_URL =
  process.env.RABBITMQ_URL ||
  `amqp://${process.env.RABBITMQ_USER || 'admin'}:${process.env.RABBITMQ_PASSWORD || 'StrongPassword123'}@${process.env.RABBITMQ_HOST || 'localhost'}:${process.env.RABBITMQ_PORT || 5672}`;

const EXCHANGE_NAME = process.env.RABBITMQ_EXCHANGE || 'examora.topic';
const ROUTING_KEY = process.env.RABBITMQ_ROUTING_KEY || 'ai.generate';
const QUEUE_NAME = process.env.RABBITMQ_QUEUE || 'ai.generation';
const DLX_NAME = process.env.RABBITMQ_DLX || 'examora.dlx';
const DLQ_NAME = process.env.RABBITMQ_DLQ || 'ai.generation.dlq';

let connection = null;
let channel = null;
let connectingPromise = null;
const returnedMessages = new Map();

async function setupTopology(currentChannel) {
  await currentChannel.assertExchange(EXCHANGE_NAME, 'topic', { durable: true });
  await currentChannel.assertExchange(DLX_NAME, 'direct', { durable: true });

  await currentChannel.assertQueue(DLQ_NAME, { durable: true });
  await currentChannel.bindQueue(DLQ_NAME, DLX_NAME, DLQ_NAME);

  await currentChannel.assertQueue(QUEUE_NAME, {
    durable: true,
    arguments: {
      'x-dead-letter-exchange': DLX_NAME,
      'x-dead-letter-routing-key': DLQ_NAME,
      'x-max-length': 1000,
      'x-overflow': 'reject-publish',
    },
  });
  await currentChannel.bindQueue(QUEUE_NAME, EXCHANGE_NAME, ROUTING_KEY);
}

async function createConnection() {
  try {
    const currentConnection = await amqp.connect(RABBITMQ_URL);
    const currentChannel = await currentConnection.createConfirmChannel();

    currentConnection.on('error', (error) => {
      console.error('[AI Publisher] Connection error:', error.message);
    });
    currentConnection.on('close', () => {
      if (connection === currentConnection) {
        connection = null;
        channel = null;
      }
      console.log('[AI Publisher] Connection closed');
    });
    currentChannel.on('error', (error) => {
      console.error('[AI Publisher] Channel error:', error.message);
    });
    currentChannel.on('close', () => {
      if (channel === currentChannel) {
        channel = null;
      }
    });
    currentChannel.on('return', (message) => {
      const messageId = message.properties.messageId;
      if (messageId) {
        returnedMessages.set(
          messageId,
          new Error(
            `RabbitMQ returned unroutable message ${messageId}: ${message.fields.replyText || 'NO_ROUTE'}`
          )
        );
      }
    });

    await setupTopology(currentChannel);

    connection = currentConnection;
    channel = currentChannel;
    console.log(
      `[AI Publisher] RabbitMQ ready: ${EXCHANGE_NAME} --${ROUTING_KEY}--> ${QUEUE_NAME}`
    );
    return currentChannel;
  } catch (error) {
    connection = null;
    channel = null;
    console.error('[AI Publisher] Failed to connect or configure topology:', error.message);
    throw error;
  }
}

async function connect() {
  if (channel) {
    return channel;
  }
  if (!connectingPromise) {
    connectingPromise = createConnection().finally(() => {
      connectingPromise = null;
    });
  }
  return connectingPromise;
}

/**
 * Publish an AI generation task and wait for broker confirmation.
 */
export async function publishAIGeneration(data) {
  const currentChannel = await connect();
  const messageId = randomUUID();
  const message = {
    request_id: data.requestId,
    task_id: data.taskId,
    trace_id: data.traceId,
  };

  currentChannel.publish(EXCHANGE_NAME, ROUTING_KEY, Buffer.from(JSON.stringify(message)), {
    persistent: true,
    mandatory: true,
    contentType: 'application/json',
    contentEncoding: 'utf-8',
    messageId,
    correlationId: data.traceId,
    type: ROUTING_KEY,
    timestamp: Date.now(),
    headers: {
      'x-service': 'AI_Generation_Service',
      'x-trace-id': data.traceId,
    },
  });
  try {
    await currentChannel.waitForConfirms();
    await new Promise((resolve) => setImmediate(resolve));
    const returnedError = returnedMessages.get(messageId);
    if (returnedError) {
      throw returnedError;
    }
  } finally {
    returnedMessages.delete(messageId);
  }

  console.log(
    `[AI Publisher] Message confirmed: id=${messageId}, request=${data.requestId}, task=${data.taskId}`
  );
  return { messageId };
}

export async function close() {
  const currentChannel = channel;
  const currentConnection = connection;
  channel = null;
  connection = null;
  returnedMessages.clear();

  try {
    if (currentChannel) {
      await currentChannel.close();
    }
    if (currentConnection) {
      await currentConnection.close();
    }
    console.log('[AI Publisher] Connection closed gracefully');
  } catch (error) {
    console.error('[AI Publisher] Error closing connection:', error.message);
  }
}

export default { publishAIGeneration, close };

File: backend/AI_Worker_Service/app/infrastructure/mq/rabbitmq.py
"""
RabbitMQ connection manager and message publisher.
Uses aio-pika for async AMQP communication.
Supports dead-letter queue (DLQ) for failed messages.
"""

import asyncio
import json
import uuid
from typing import Any

import aio_pika
from aio_pika import DeliveryMode, ExchangeType, Message
from aio_pika.abc import AbstractChannel, AbstractConnection, AbstractQueue

from app.core.config import get_settings
from app.core.exceptions import MessageQueueError
from app.core.logging import get_logger

logger = get_logger(__name__)

_connection: AbstractConnection | None = None
_channel: AbstractChannel | None = None
_queues_setup_for_channel_id: int | None = None
_queues_setup_lock = asyncio.Lock()


async def get_rabbitmq_connection() -> AbstractConnection:
    """Return a singleton RabbitMQ connection."""
    global _connection
    if _connection is None or _connection.is_closed:
        settings = get_settings()
        try:
            _connection = await aio_pika.connect_robust(
                settings.rabbitmq_url,
                reconnect_interval=settings.rabbitmq_reconnect_interval_seconds,
                fail_fast=False,
            )
            logger.info("RabbitMQ connected: %s", settings.rabbitmq_url.split("@")[-1])
        except Exception as e:
            raise MessageQueueError(f"Cannot connect to RabbitMQ: {e}") from e
    return _connection


async def get_rabbitmq_channel() -> AbstractChannel:
    """Return a channel from the singleton connection."""
    global _channel
    conn = await get_rabbitmq_connection()
    if _channel is None or _channel.is_closed:
        settings = get_settings()
        _channel = await conn.channel()
        await _channel.set_qos(prefetch_count=settings.rabbitmq_prefetch_count)
    return _channel


async def setup_queues(channel: AbstractChannel) -> tuple[AbstractQueue, AbstractQueue]:
    """
    Declare the AI topic exchange, main queue, DLX, DLQ, and bindings.

    Returns:
        (main_queue, dlq)
    """
    settings = get_settings()

    main_exchange = await channel.declare_exchange(
        settings.rabbitmq_exchange,
        ExchangeType.TOPIC,
        durable=True,
    )
    dlx = await channel.declare_exchange(
        settings.rabbitmq_dlx,
        ExchangeType.DIRECT,
        durable=True,
    )

    # The DLQ has no dead-letter routing of its own.
    dlq = await channel.declare_queue(
        settings.rabbitmq_dlq,
        durable=True,
        arguments={},
    )
    await dlq.bind(dlx, routing_key=settings.rabbitmq_dlq)

    # Failed messages are routed through the DLX into the DLQ.
    # x-max-length protects RabbitMQ from unbounded memory growth.
    # x-overflow=reject-publish rejects new messages when max is reached.
    main_queue = await channel.declare_queue(
        settings.rabbitmq_queue,
        durable=True,
        arguments={
            "x-dead-letter-exchange": settings.rabbitmq_dlx,
            "x-dead-letter-routing-key": settings.rabbitmq_dlq,
            "x-max-length": settings.rabbitmq_queue_max_length,
            "x-overflow": settings.rabbitmq_overflow_policy,
        },
    )
    await main_queue.bind(
        main_exchange,
        routing_key=settings.rabbitmq_routing_key,
    )

    logger.info(
        "RabbitMQ topology ready | exchange=%s routing_key=%s queue=%s "
        "dlx=%s dlq=%s",
        settings.rabbitmq_exchange,
        settings.rabbitmq_routing_key,
        settings.rabbitmq_queue,
        settings.rabbitmq_dlx,
        settings.rabbitmq_dlq,
    )
    return main_queue, dlq


async def ensure_queues_setup(channel: AbstractChannel) -> None:
    """Ensure queues/exchanges are declared once per channel lifecycle."""
    global _queues_setup_for_channel_id
    channel_id = id(channel)

    if _queues_setup_for_channel_id == channel_id:
        return

    async with _queues_setup_lock:
        if _queues_setup_for_channel_id == channel_id:
            return
        await setup_queues(channel)
        _queues_setup_for_channel_id = channel_id


async def publish_task_message(
    request_id: str,
    task_id: str,
    trace_id: str,
    extra_payload: dict[str, Any] | None = None,
) -> None:
    """
    Publish a generation task message to the main queue.
    Message is persistent (survives broker restart).
    """
    settings = get_settings()
    try:
        channel = await get_rabbitmq_channel()
        await ensure_queues_setup(channel)

        payload: dict[str, Any] = {
            "request_id": request_id,
            "task_id": task_id,
            "trace_id": trace_id,
        }
        if extra_payload:
            payload.update(extra_payload)

        message = Message(
            body=json.dumps(payload).encode(),
            delivery_mode=DeliveryMode.PERSISTENT,
            message_id=str(uuid.uuid4()),
            content_type="application/json",
        )

        exchange = await channel.get_exchange(settings.rabbitmq_exchange)
        await exchange.publish(
            message,
            routing_key=settings.rabbitmq_routing_key,
        )
        logger.info(
            "Published task message | exchange=%s routing_key=%s request=%s task=%s",
            settings.rabbitmq_exchange,
            settings.rabbitmq_routing_key,
            request_id,
            task_id,
        )
    except MessageQueueError:
        raise
    except Exception as e:
        raise MessageQueueError(f"Failed to publish message: {e}") from e


async def close_rabbitmq() -> None:
    """Gracefully close RabbitMQ connection."""
    global _connection, _channel, _queues_setup_for_channel_id
    if _channel and not _channel.is_closed:
        await _channel.close()
        _channel = None
    if _connection and not _connection.is_closed:
        await _connection.close()
        _connection = None
    _queues_setup_for_channel_id = None
    logger.info("RabbitMQ connection closed.")
